File: Repositories/python/python-monolith-images/src/similarity.py
import math
from src.data_load_module import feature_index


def euclidean(a, b):
    s = 0.0
    for x, y in zip(a, b):
        d = x - y
        s += d * d
    return math.sqrt(s)


def find_similar_images(query_vector, threshold):
    result = []
    for image_id, ref in feature_index.items():
        dist = euclidean(query_vector, ref)
        print(f"[SIM] {image_id} dist={dist:.4f}")
        if dist <= threshold:
            result.append(image_id)
    return result

File: Repositories/js/js-monolith-images/similarity.js
import { featureIndex } from "./dataLoadModule.js";

function euclidean(a, b) {
    let sum = 0;
    for (let i = 0; i < a.length; i++) {
        const d = a[i] - b[i];
        sum += d * d;
    }
    return Math.sqrt(sum);
}

export function findSimilarImages(queryVector, threshold) {

    return Object.entries(featureIndex)
        .filter(([id, ref]) => {
            const distance = euclidean(queryVector, ref);

            console.log(`[SIM] ${id} dist=${distance.toFixed(4)}`);

            return distance <= threshold;
        })
        .map(([id]) => id);
}

